### PR TITLE
JIT ARM64/SVE2: Add increment/decrement from predicate register instructions

### PR DESCRIPTION
**Pull request by amanasifkhalid**
This implements the following instructions:
* `incp/decp <Xdn>, <Pm>.<T>`: increment/decrement general register by predicate count
* `incp/decp <Zdn>.<T>, <Pm>.<T>`: increment/decrement vector register by predicate count
* `sqdecp/sqincp <Xdn>, <Pm>.<T>, <Wdn>`: signed saturating increment/decrement general register by predicate count
* `sqdecp/sqincp <Zdn>.<T>, <Pm>.<T>`: signed saturating increment/decrement vector register by predicate count
* `uqdecp/uqincp <Wdn>, <Pm>.<T>`: unsigned saturating increment/decrement general register by predicate count
* `uqdecp/uqincp <Zdn>.<T>, <Pm>.<T>`: unsigned saturating increment/decrement vector register by predicate count

### CODE REVIEW

File: src/coreclr/jit/codegenarm64.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenarm64.cpp:6367]: Temporarily we should probably add all the predicate registers to registerarm64.h but define them as equal to the vector register numbers, e.g.:
```
REGALIAS(REG_P0, REG_V0)
REGALIAS(REG_P1, REG_V1)
...
```

Then use `REG_P1` here instead of `REG_R1`

File: src/coreclr/jit/codegenarm64.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenarm64.cpp:6368]: If this takes a `Z` register than it should be `REG_V0` instead of `REG_R0`?

File: src/coreclr/jit/emitarm64.cpp
Comment (BruceForstall) [inline src/coreclr/jit/emitarm64.cpp:959]: Should these forms be merged?

(Or should we wait to decide on merging until after many/all instructions have been implemented?)

File: src/coreclr/jit/codegenarm64.cpp
Comment (BruceForstall) [reply src/coreclr/jit/codegenarm64.cpp:6367]: `emitter::isPredicateRegister` could use these definitions.

File: src/coreclr/jit/emitarm64.cpp
Comment (BruceForstall) [inline src/coreclr/jit/emitarm64.cpp:13928]: Seems like we need a `emitPredicateRegName` than handles the "fake" predicate registers but still outputs `p0` through `p15`

File: src/coreclr/jit/emitarm64.h
Comment (BruceForstall) [inline src/coreclr/jit/emitarm64.h:674]: This could use the new REG_P0 to REG_P15. And/or define `REG_P_FIRST` and `REG_P_LAST`?

File: src/coreclr/jit/codegenarm64.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/codegenarm64.cpp:6368]: Yes, thanks for catching that.

File: src/coreclr/jit/codegenarm64.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/codegenarm64.cpp:6367]: I think that's a good idea. Kunal has a couple of similar TODOs in #95016, so if this PR gets merged in first, maybe we should update his PR to use the aliases. @kunalspathak

File: src/coreclr/jit/emitarm64.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/emitarm64.cpp:959]: I'm ok with either doing it now or waiting until after we've implemented all of them. @kunalspathak @TIHan do you have any preference for this?

File: src/coreclr/jit/emitarm64.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/emitarm64.cpp:13928]: Kunal has a temporary implementation for this in #95016. As an aside, there seems to be potential for circular dependencies between these two PRs. Should I wait for Kunal's to be merged in first? I can then update my implementation with the new methods #95016 adds in (like `emitPredicateRegName`) and also update the predicate register TODOs to use the aliases you suggested above.

**Comment by kunalspathak** [comment]
@amanasifkhalid - I will update https://github.com/dotnet/runtime/pull/95016 to add the following:
- Add `emitDispSveRegister()` to display "Z" registers
- Add fake predicate registers and corresponding methods (thanks @BruceForstall for the suggestion)

Once #95016 along with these changes are merged, it will become easier to base the changes on top of it. We still won't be able to merge it because of encoding validation.

File: src/coreclr/jit/codegenarm64.cpp
Comment (kunalspathak) [reply src/coreclr/jit/codegenarm64.cpp:6368]: Done. Please use the new files in https://github.com/dotnet/runtime/issues/94549

File: src/coreclr/jit/emitarm64.cpp
Comment (kunalspathak) [reply src/coreclr/jit/emitarm64.cpp:959]: > waiting until after we've implemented all of them

Will be good to wait so we can track the coverage of them as listed in https://github.com/dotnet/runtime/issues/94549. We can then have one pass to combine all the duplicate ones.

File: src/coreclr/jit/emitarm64.h
Comment (kunalspathak) [reply src/coreclr/jit/emitarm64.h:674]: This is done in https://github.com/dotnet/runtime/pull/95016

**Comment by amanasifkhalid** [comment]
I merged in Kunal's changes from #95016 and fixed the unit test additions. I'm waiting for #95105 to be merged in before using the updated machine-generated files, since the new `emitOutputInstr` code relies on the `insEncodeReg_*` methods.

> We still won't be able to merge it because of encoding validation.

Would it make sense for me to open more (draft) PRs like this one if we will have to wait to merge them? I guess this is a good way to find issues with the tool-generated code.

**Comment by kunalspathak** [comment]
@amanasifkhalid - I have shared few steps offline on validating the encoding, see if you can check them. Please hold off on doing more draft PRs as I am updating the tool to fix few things that came up in Alan's #95127 . You might want to consume the new files.

**Comment by BruceForstall** [comment]
Can you update the comment at the top? (If nothing else, the "once that work is merged in, I'll rebase these changes on top of it" comment is obsolete)

**Comment by BruceForstall** [comment]
cc @a74nh

File: src/coreclr/jit/codegenarm64test.cpp
Comment (a74nh) [inline src/coreclr/jit/codegenarm64test.cpp:5102]: For consistency with instructions elsewhere, INS_OPTS_SCALABLE_B should be INS_OPTS_SCALABLE_B_WITH_SCALAR. Same for anything else where EA_ is a value instead of  EA_SCALABLE.

File: src/coreclr/jit/emitarm64.cpp
Comment (a74nh) [inline src/coreclr/jit/emitarm64.cpp:1279]: Why might these two groups write to GC Reg? (It's possible I've been missing some from my PRs)

File: src/coreclr/jit/emitarm64.cpp
Comment (a74nh) [inline src/coreclr/jit/emitarm64.cpp:17227]: Once https://github.com/dotnet/runtime/pull/95918 goes in then this should be 
`emitDispPredicateReg(id->idReg2(), PREDICATE_SIZED, false);` and then don't call `emitDispArrangement()`

(This code can be fixed up later if it goes in before 95918)

File: src/coreclr/jit/emitarm64.cpp
Comment (a74nh) [inline src/coreclr/jit/emitarm64.cpp:15455]: I don't see any use of calling `emitDispArrangement()` with `addComma` set to true.

**Review by a74nh — COMMENTED** [review]
Just a few small things

File: src/coreclr/jit/emitarm64.cpp
Comment (kunalspathak) [reply src/coreclr/jit/emitarm64.cpp:15455]: It is used at https://github.com/dotnet/runtime/pull/94811/files#diff-2b2c8b9011607926410624d6f81613fad7b74c6e0516d578675a8b792998fe4fR17249

Comment (a74nh) [reply src/coreclr/jit/emitarm64.cpp:15455]: Thanks. (But still can be removed if end up using PREDICATE_SIZED)

File: src/coreclr/jit/emitarm64.cpp
Comment (amanasifkhalid) [reply src/coreclr/jit/emitarm64.cpp:1290]: Ah sorry, didn't mean to push that in the first place. I don't think we should worry about touching `emitInsMayWriteToGCReg` for now. Once we start emitting these SVE instructions for actual codegen, we might need to come back to this method.

**Comment by amanasifkhalid** [comment]
Thank you for the review @a74nh! I've applied your feedback, and the unit test outputs remain the same.

JitDisasm:
```
decp    x0, p0.b
decp    x1, p1.h
decp    x2, p2.s
decp    x3, p3.d
incp    x4, p4.b
incp    x5, p5.h
incp    x6, p6.s
incp    x7, p7.d
decp    z0.h, p0.h
decp    z1.s, p1.s
decp    z2.d, p2.d
incp    z3.h, p3.h
incp    z4.s, p4.s
incp    z5.d, p5.d
sqdecp  x0, p0.b, w0
sqdecp  x1, p1.h, w1
sqdecp  x2, p2.s, w2
sqdecp  x3, p3.d, w3
sqdecp  x4, p4.b
sqdecp  x5, p5.h
sqdecp  x6, p6.s
sqdecp  x7, p7.d
sqincp  x0, p0.h, w0
sqincp  x1, p1.s, w1
sqincp  x2, p2.b, w2
sqincp  x3, p3.d, w3
sqincp  x4, p4.b
sqincp  x5, p5.h
sqincp  x6, p6.s
sqincp  x7, p7.d
uqdecp  w0, p0.b
uqdecp  w1, p1.h
uqdecp  w2, p2.s
uqdecp  w3, p3.d
uqdecp  x4, p4.b
uqdecp  x5, p5.h
uqdecp  x6, p6.s
uqdecp  x7, p7.d
uqincp  w0, p0.b
uqincp  w1, p1.h
uqincp  w2, p2.s
uqincp  w3, p3.d
uqincp  x4, p4.b
uqincp  x5, p5.h
uqincp  x6, p6.s
uqincp  x7, p7.d
sqdecp  z0.h, p0.h
sqdecp  z1.s, p1.s
sqdecp  z2.d, p2.d
sqincp  z3.h, p3.h
sqincp  z4.s, p4.s
sqincp  z5.d, p5.d
uqdecp  z6.h, p6.h
uqdecp  z7.s, p7.s
uqdecp  z8.d, p0.d
uqincp  z9.h, p1.h
uqincp  z10.s, p2.s
uqincp  z11.d, p3.d
```

cstool:
```
00882D25  decp  x0, p0.b
21886D25  decp  x1, p1.h
4288AD25  decp  x2, p2.s
6388ED25  decp  x3, p3.d
84882C25  incp  x4, p4.b
A5886C25  incp  x5, p5.h
C688AC25  incp  x6, p6.s
E788EC25  incp  x7, p7.d
00806D25  decp  z0.h, p0.h
2180AD25  decp  z1.s, p1.s
4280ED25  decp  z2.d, p2.d
63806C25  incp  z3.h, p3.h
8480AC25  incp  z4.s, p4.s
A580EC25  incp  z5.d, p5.d
00882A25  sqdecp        x0, p0.b, w0
21886A25  sqdecp        x1, p1.h, w1
4288AA25  sqdecp        x2, p2.s, w2
6388EA25  sqdecp        x3, p3.d, w3
848C2A25  sqdecp        x4, p4.b
A58C6A25  sqdecp        x5, p5.h
C68CAA25  sqdecp        x6, p6.s
E78CEA25  sqdecp        x7, p7.d
00886825  sqincp        x0, p0.h, w0
2188A825  sqincp        x1, p1.s, w1
42882825  sqincp        x2, p2.b, w2
6388E825  sqincp        x3, p3.d, w3
848C2825  sqincp        x4, p4.b
A58C6825  sqincp        x5, p5.h
C68CA825  sqincp        x6, p6.s
E78CE825  sqincp        x7, p7.d
00882B25  uqdecp        w0, p0.b
21886B25  uqdecp        w1, p1.h
4288AB25  uqdecp        w2, p2.s
6388EB25  uqdecp        w3, p3.d
848C2B25  uqdecp        x4, p4.b
A58C6B25  uqdecp        x5, p5.h
C68CAB25  uqdecp        x6, p6.s
E78CEB25  uqdecp        x7, p7.d
00882925  uqincp        w0, p0.b
21886925  uqincp        w1, p1.h
4288A925  uqincp        w2, p2.s
6388E925  uqincp        w3, p3.d
848C2925  uqincp        x4, p4.b
A58C6925  uqincp        x5, p5.h
C68CA925  uqincp        x6, p6.s
E78CE925  uqincp        x7, p7.d
00806A25  sqdecp        z0.h, p0.h
2180AA25  sqdecp        z1.s, p1.s
4280EA25  sqdecp        z2.d, p2.d
63806825  sqincp        z3.h, p3.h
8480A825  sqincp        z4.s, p4.s
A580E825  sqincp        z5.d, p5.d
C6806B25  uqdecp        z6.h, p6.h
E780AB25  uqdecp        z7.s, p7.s
0880EB25  uqdecp        z8.d, p0.d
29806925  uqincp        z9.h, p1.h
4A80A925  uqincp        z10.s, p2.s
6B80E925  uqincp        z11.d, p3.d
```

File: src/coreclr/jit/emitarm64.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64.cpp:14881]: technically, you could combine this with `IF_SVE_DM_2A`, but lets do it in cleanup phase.

```c++
case IF_SVE_DO_2A:
  code |= insEncodeVLSElemsize(id->idOpSize());                    // X
  FALLTHROUGH;

case IF_SVE_DM_2A:
  ...
```

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Comment by amanasifkhalid** [comment]
Merged changes from #95996.

**Pull request merged**
